File: 0200-number-of-islands/0200-number-of-islands.java
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int numIslands(char[][] grid) {
        if (grid == null) {
            return 0;
        }

        int islands = 0;
        int ROWS = grid.length;
        int COLS = grid[0].length;

        for (int r = 0; r < ROWS; r++) {
            for (int c = 0; c < COLS; c++) {
                if (grid[r][c] == '1') {
                    islands += 1;
                    bfs(grid, r, c);
                }
            }
        }

        return islands;
    }

    public boolean isValid(int r, int c, char[][] grid) {
        int ROWS = grid.length;
        int COLS = grid[0].length;

        return r >= 0 && r != ROWS && c >= 0 && c != COLS && grid[r][c] != '0';
    }

    public void bfs(char[][] grid, int r, int c) {

        Queue<int[]> queue = new LinkedList<>();
        grid[r][c] = '0';

        queue.add(new int[]{r, c});

        while (!queue.isEmpty()) {
            int[] cell = queue.poll();
            int row = cell[0];
            int col = cell[1];

            int[][] neighbours = new int[][]{new int[]{row + 1, col}, new int[]{row - 1, col}, new int[]{row, col + 1}, new int[]{row, col - 1}};

            for (int[] neighbour: neighbours) {
                int newRow = neighbour[0];
                int newCol = neighbour[1];

                if (isValid(newRow, newCol, grid)) {
                    queue.add(new int[]{newRow, newCol});
                    grid[newRow][newCol] = '0';
                }
            }
        }
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int: 
        
        if not grid:
            return 0 
        
        ROWS, COLS = len(grid), len(grid[0])
        visit = set() 
        number = 0 
        
        def isValid(r, c):
            
            if (r < 0 or r == ROWS or c < 0 or c == COLS) or (grid[r][c] == "0") or ((r, c) in visit):
                return False 

            return True 
                        
        def bfs(r, c):
            queue = [] 
            queue.append((r, c))
            visit.add((r, c))
            
            while queue:
                r, c = queue.pop(0) 
                neighbours = [(r + 1, c), (r - 1, c), (r, c + 1), (r, c - 1)]
                
                for nx, ny in neighbours:
                    if isValid(nx, ny):
                        visit.add((nx, ny))
                        queue.append((nx, ny))
                        

        for r in range(0, ROWS):
            for c in range(0, COLS): 
                if grid[r][c] == "1" and (r, c) not in visit:
                    number += 1
                    bfs(r, c)
                    
        return number 
                
            
            
        
        
            
        
        
        
        
        